pkg/controller/sparkapplication/job.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2017 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sparkapplication
+
+import (
+	"fmt"
+	"strings"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1alpha1"
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/config"
+)
+
+type submissionJobManager struct {
+	kubeClient kubernetes.Interface
+}
+
+func (sjm *submissionJobManager) createSubmissionJob(s *submission) (*batchv1.Job, error) {
+	var image string
+	if s.app.Spec.Image != nil {
+		image = *s.app.Spec.Image
+	} else if s.app.Spec.Driver.Image != nil {
```

Comment:
As we discussed, using customer image to run spark-submit might be a problem given that Job controller could start multiple pods running spark-submit.

Should we use our own images with a wrapper script over spark-submit ?

Reply:
It's the Job's lack of exactly-once semantics that might lead to that issue, not which image is used. Since we set the name of the Kubernetes Job deterministically, this should not be an issue.

Reply:
Yes, but from my understanding, if the Job ends up creating 2 pods, one of the pod will fail (because spark-submit will fail due to an AlreadyExists error from the kube-api) and that can potentially cause the Job to fail leading to the Operator marking the SparkApplication as Failed when there was a successful spark-submit.

As we discussed in our sync-up, If we use custom spark distribution images with a wrapper over spark-submit, we can identify these cases (when we get an AlreadyExists error from kube-api) and mark the pod as Successful to prevent the issue explained above.

Reply:
@liyinan926 i think I agree with you both, which is weird. I see your point that running customer images would relieve us from the pressure of managing our own distribution images. But, failure in the job execution (which we have seen a few times) due to non-idempotent behavior is confusing for the user. It will essentially result in the Application (Job) failing, when it reality it was executed (or is in progress). This has repercussions on 
- **correctness**: user expect the job to not have started/executed, but it has infact executed and may affect a retry - _like corrupted data_
- **cost attribution**: user expects that the job is not running, but the job executes and ends up used unintended resources. This can be extremely important for companies that run on cloud and every extra compute hour spent affects its bottom line.

Its a hard choice, but I will choose _correctness_ and _predictable cost_ over simplified user interface. Also in this case both the options do not really affect the user experience.

Reply:
On talking more with Anmol, _Pods_ may have the same problem as the _Jobs_ and both the above mentioned problems are true with this too.
the real solution seems to be to make `spark-submit` more reliable and idempotent. Each run of spark-submit should result in success if the pod with that name exists or a predictable exit code.
I thnk predictable exit-code is a better solution and we will handle that in the operator.

On that note, I am ok with you going ahead with this change and hopefully asking Spark to make a change to spark-submit?

Reply:
Also can you document this as part of the PR, and probably we can open an issue with spark and link that in the docs,